### PR TITLE
PR1: add support for accounting export and expense creations

### PR DESCRIPTION
**Pull request by NileshPant1999**

### CODE REVIEW

File: apps/fyle/models.py [hunk 1]
```diff
@@ -104,6 +112,71 @@ class Expense(BaseModel):
...
+                    'verified_at': expense['verified_at'],
+                    'custom_properties': expense['custom_properties'],
+                    'payment_number': expense['payment_number'],
+                    'file_ids': expense['file_ids'],
+                    'corporate_card_id': expense['corporate_card_id'],
+                }
+            )
+
+            # Doing this to avoid circular import
+            AccountingExport = import_string('apps.accounting_exports.models.AccountingExport')
```
Comment (NileshPant1999) [inline apps/fyle/models.py:172]: this is to avoid circular import

File: apps/fyle/models.py [hunk 2]
```diff
@@ -104,6 +112,71 @@ class Expense(BaseModel):
...
+
+            # Check if an AccountingExport related to the expense object already exists
+            if not AccountingExport.objects.filter(expenses__id=expense_object.id).first():
```
Comment (Shwetabhk) [inline apps/fyle/models.py:175]:
```python
expenses_with_accounting_export = Expense.objects.filter(accountingexport__isnull=False).distinct()
```

Try something like this

File: apps/accounting_exports/models.py [hunk 3]
```diff
@@ -50,6 +100,56 @@ class AccountingExport(BaseForeignWorkspaceModel):
     class Meta:
         db_table = 'accounting_exports'
 
+    @staticmethod
+    def create_accounting_export_report_id(expense_objects: List[Expense], fund_source: str, workspace_id):
```
Comment (Shwetabhk) [inline apps/accounting_exports/models.py:104]:
```suggestion
    def create_accounting_export(expense_objects: List[Expense], fund_source: str, workspace_id):
```

File: apps/accounting_exports/models.py [hunk 4]
```diff
@@ -30,6 +41,45 @@
...
+    Group expenses based on specified fields
+    """
+
+    credit_card_expense_grouped_by = export_setting.credit_card_expense_grouped_by
+    credit_card_expense_date = export_setting.credit_card_expense_date
+    reimbursable_expense_grouped_by = export_setting.reimbursable_expense_grouped_by
+    reimbursable_expense_date = export_setting.reimbursable_expense_date
+
+    if fund_source == 'CCC':
+        group_fields = ['report_id', 'claim_number'] if credit_card_expense_grouped_by == 'REPORT' else ['expense_id', 'expense_number']
```
Comment (Shwetabhk) [inline apps/accounting_exports/models.py:61]:
```python
{
    'CCC': {
        'REPORT': [..],
        'EXPENSE': [..]
    }
}

File: apps/accounting_exports/models.py [hunk 5]
```diff
@@ -30,6 +41,45 @@
...
+        if credit_card_expense_date != 'LAST_SPENT_AT':
+            group_fields.append(credit_card_expense_date.lower())
+
+    if fund_source == 'PERSONAL':
```
Comment (Shwetabhk) [inline apps/accounting_exports/models.py:65]: the group fields are same, you can merge the two condition

File: apps/accounting_exports/models.py [hunk 6]
```diff
@@ -30,6 +41,45 @@
...
+        group_fields = ['report_id', 'claim_number'] if reimbursable_expense_grouped_by == 'REPORT' else ['expense_id', 'expense_number']
+        if reimbursable_expense_date != 'LAST_SPENT_AT':
+            group_fields.append(reimbursable_expense_date.lower())
+
+    # Extract expense IDs from the provided expenses
+    expense_ids = [expense.id for expense in expenses]
+
```
Comment (Shwetabhk) [inline apps/accounting_exports/models.py:72]: add something like
```python
    group_fields.append(fund_source)
```

File: apps/accounting_exports/models.py [hunk 7]
```diff
@@ -50,6 +100,56 @@ class AccountingExport(BaseForeignWorkspaceModel):
...
+                expense_group['last_spent_at'] = latest_expense.spent_at if latest_expense else None
+
+            # Store expense IDs and remove unnecessary keys
+            expense_ids = expense_group['expense_ids']
+            expense_group.pop('total')
+            expense_group.pop('expense_ids')
+
+            # Format date fields according to the specified format
+            for key in expense_group:
+                if key in ALLOWED_FORM_INPUT['export_date_type']:
```
Comment (Shwetabhk) [inline apps/accounting_exports/models.py:137]: ALLOWED_FORM_INPUT and all is not required

File: apps/accounting_exports/models.py [hunk 8]
```diff
@@ -50,6 +100,56 @@ class AccountingExport(BaseForeignWorkspaceModel):
...
+        """
+
+        # Retrieve the ExportSetting for the workspace
+        export_setting = ExportSetting.objects.get(workspace_id=workspace_id)
+
+        # Initialize lists and fields for reimbursable and corporate credit card expenses
+        expense_group_fields = ['employee_email', 'fund_source']
+
+        # Group expenses based on specified fields and fund_source
+        expense_groups = _group_expenses(expense_objects, expense_group_fields, export_setting, fund_source)
```
Comment (Shwetabhk) [inline apps/accounting_exports/models.py:116]: user term `accounting_exports` everywhere
remove `expense_groups` completely

File: apps/accounting_exports/models.py [hunk 9]
```diff
@@ -50,6 +100,56 @@ class AccountingExport(BaseForeignWorkspaceModel):
...
+
+        for expense_group in expense_groups:
+            # Determine the date field based on fund_source
+            if fund_source == 'PERSONAL':
```
Comment (Shwetabhk) [inline apps/accounting_exports/models.py:120]: the conditions can be cleaned up using dictionaties

Comment (Shwetabhk) [reply apps/accounting_exports/models.py:120]: data = EXPORT_SETTINGS_CONFIG[fund_source]['date_field']

**Review by Shwetabhk — CHANGES_REQUESTED** [review]

**Review by Shwetabhk — APPROVED** [review]

**Pull request merged**